### DISCUSSION TITLE
Visual Studio 2019 IntelliSense really slow

### DISCUSSION THREAD

**Discussion by Slion** [microsoft/CsWin32#591] [Q&A] [answered]
Is it just me or IntelliSense does not cope well when generating entire modules.
In my case I had the following listed in `NativeMethods.txt` :

```
Hid.*
setupapi.*
Kernel32.*
User32.*
```

Makes typing any code frustratingly slow at times.
Would it be better if I moved all those exports in a dedicated NuGet package? Don't you guys already publish such NuGet packages?

Visual Studio 2022 seems to deal with that a bit better but not by far.

**Comment by AArnott** [comment, marked as answer]
Perf is an ongoing concern and one that we can certainly improve on. If you can use Report a Problem inside VS and indicate you have a perf concern and let it 'watch' you while you repro it, it will collect a perf trace so we can see just what's making it slow, to inform our efforts at improving performance. Be sure to drop a link to your feedback ticket here so I can ensure it gets routed to the right place. 

Pre-built NuGet packages is something we're considering. In fact the https://github.com/dotnet/pinvoke repo may be replaced with code generated by CsWin32 eventually. But in the meantime, my recommendation is to _not_ generate more APIs than you need. Whole modules is making your life slower than is likely necessary.

**Comment by Gavin-Williams** [comment]
I noticed this straight away when I started using CsWin32. Terrible performance. I've just upgraded my computer and it feels like Visual Studio is running on a single thread. Character display lags behind typing, some Intellisense operations take many seconds to complete. And I'm not using whole namespace imports. I'm using explicit method imports.

**Reply by Slion** [reply]
Don't think I've had any problems since I stopped doing full module import.

**Reply by AArnott** [reply]
@gavin-williams please see [my answer](https://github.com/microsoft/CsWin32/discussions/591#discussioncomment-2986297), specifically about collecting a perf trace so we can analyze the problem and hopefully fix it.

**Reply by Gavin-Williams** [reply]
I'll see if I can capture the behavior. I've pulled the small CsWin32 project out by itself - just to see how it behaves in that situation - it was in a solution with a few other projects before. I guess I was wondering if maybe there was something going on because of my other projects. I can't see that being the case. But anyway it's worth trying out. So give me some time to work with the project in a simple solution and my main solution so I can see what's happening.

**Reply by Gavin-Williams** [reply]
Here's a quick capture of the issue. I have experienced much worse lag than this though, this was pretty tolerable. I will capture a better example, but this might give some evidence of something going on...
https://developercommunity.visualstudio.com/t/CsWin32-possibly-causing-lag-in-VS/10100351

**Reply by AArnott** [reply]
Great. I see 14 seconds of CPU time spent in the CsWin32 source generator in that trace, so I'll see where the time is going and whether we can improve on that.

**Reply by Gavin-Williams** [reply]
Sam Harwell @ MS says it's Semantic Colorizer affecting my experience. I've uninstalled the offending extension and reported the issue to the author. I see that someone else has also reported the issue with that product just a few days ago. Not sure if that changes anything in regards to what you've observed. Anyway, I'll see how VS goes over the next few days and report back.

**Reply by AArnott** [reply]
Some results:

Milliseconds | Use
--|--
1600 | Computing suggestions because something in NativeMethods.txt doesn't match any API. When you compile, do you see errors or warnings about this? Cleaning this up should speed things up a bit. We use fuzzy matching to make suggestions, which is expensive.
1600 | Calling `PEReaderExtensions.GetMetadataReader`. We do this repeatedly and evidently it's an expensive operation. We might pool the result of this call instead of calling it for every code generation request.
1056 | `GetMethodsByName` spends 472ms concatenating strings and another 399ms in 3 dictionary lookups *per namespace* in the search. We may be able to consolidate this to just 1 dictionary lookup with no string concatenation.
650 | `Generator.TryGetEnumName`. I have a change to make this dramatically faster, but as this tends to execute when NativeMethods.txt contains something that ultimately will produce a warning or error, you could improve your experience now by resolving that.
511 | `SyntaxNode.GetText`, which renders the syntax tree to a `SourceText`. Nothing we can really do about that until/unless we start caching and reusing `SourceText`, which is the ultimate in source generator caching and if implemented could save us much more time than just this.
183 | `TryGenerateConstant` spends this much time in direct dictionary lookups (one lookup per namespace searched). We might be able to get this down to just one lookup.
183 | `TryGenerateType` spends this much time in direct dictionary lookups (one lookup per namespace searched). We might be able to get this down to just one lookup.

That still leaves 9.8 seconds, so we're talking only incremental improvement here. Unless as part of the SourceText improvement we actually get to caching the `SourceText`, which would be a night-and-day improvement. I'll make some improvements today or file issues and add them to the table above to track those individual efforts.

**Reply by AArnott** [reply]
FYI @sharwell

**Reply by sharwell** [reply]
@AArnott we intentionally bypass the source generator execution on the inner typing loop (we assume the previous execution was correct and don't attempt to validate or update this assumption). The extension was overriding this fast path and causing the source generator to run. The fact that the source generator was running _at all_ for classification indicated a bug.

**Reply by AArnott** [reply]
Good to know. @sharwell Do you think any of the items in my table are still worthwhile to fix? If we're talking an order of magnitude in excess SG executions, then I guess most/all these items become uninteresting.

**Reply by AArnott** [reply]
Perf work in the above table is now tracked in https://github.com/microsoft/CsWin32/issues/612.
